src/components/tl/mlx5/alltoall/alltoall.c

```diff
@@ -0,0 +1,751 @@
+    }
+
+    topo      = team->topo;
+    node      = ucc_topo_get_sbgp(topo, UCC_SBGP_NODE);
+    net       = ucc_topo_get_sbgp(topo, UCC_SBGP_NODE_LEADERS);
+    node_size = node->group_size;
+    nnodes    = ucc_topo_nnodes(topo);
+    team_size = UCC_TL_TEAM_SIZE(team);
+
+    if (!ucc_topo_isoppn(topo)) {
+        tl_debug(ctx->super.super.lib,
+                 "disabling mlx5 a2a for team with non-uniform ppn, "
+                 "min_ppn %d, max_ppn %d",
+                 ucc_topo_min_ppn(topo), ucc_topo_max_ppn(topo));
+        status = UCC_ERR_NOT_SUPPORTED;
+        goto err;
+    }
+    ppn = ucc_topo_max_ppn(topo);
+
+    if (net->status == UCC_SBGP_NOT_EXISTS) {
+        tl_debug(ctx->super.super.lib,
+                 "disabling mlx5 a2a for single node team");
+        status = UCC_ERR_NOT_SUPPORTED;
+        goto err;
+    }
+
+    if (nnodes == team_size) {
+        tl_debug(ctx->super.super.lib,
+                 "disabling mlx5 a2a for ppn=1 case, not supported so far");
+        status = UCC_ERR_NOT_SUPPORTED;
+        goto err;
+    }
+
+    a2a->node_size = node_size;
+    ucc_assert(team_size == ppn * nnodes);
+
+    for (i = 0; i < nnodes; i++) {
+        for (j = 1; j < ppn; j++) {
+            if (!ucc_team_ranks_on_same_node(i * ppn, i * ppn + j,
+                                             UCC_TL_CORE_TEAM(team))) {
+                tl_debug(ctx->super.super.lib,
+                         "disabling mlx5 a2a for team with non contiguous "
+                         "ranks-per-node placement");
+                status = UCC_ERR_NOT_SUPPORTED;
+                goto err;
+            }
+        }
+    }
+
+    a2a->pd                 = ctx->shared_pd;
+    a2a->ctx                = ctx->shared_ctx;
+    a2a->ib_port            = ctx->ib_port;
+    a2a->node.sbgp          = node;
+    a2a->net.sbgp           = net;
+    a2a->node.asr_rank      = MLX5_ASR_RANK;
+    a2a->num_dci_qps        = UCC_TL_MLX5_TEAM_LIB(team)->cfg.num_dci_qps;
+    a2a->sequence_number    = 1;
+    a2a->net.ctrl_mr        = NULL;
+    a2a->net.remote_ctrl    = NULL;
+    a2a->net.rank_map       = NULL;
+    a2a->max_msg_size       = MAX_MSG_SIZE;
+    a2a->max_num_of_columns =
+        ucc_div_round_up(node->group_size, 2 /* todo: there can be an estimation of
+                                                    minimal possible block size */);
+
+    ucc_assert(a2a->net.sbgp->status == UCC_SBGP_ENABLED ||
+               node->group_rank != 0);
+
+    if (a2a->node.asr_rank == node->group_rank) {
+        a2a->net.net_size = a2a->net.sbgp->group_size;
+        storage_size      = OP_SEGMENT_SIZE(a2a) * MAX_OUTSTANDING_OPS;
+        a2a->bcast_data.shmid =
+            shmget(IPC_PRIVATE, storage_size, IPC_CREAT | 0600);
+        if (a2a->bcast_data.shmid == -1) {
+            tl_error(ctx->super.super.lib,
+                     "failed to allocate sysv shm segment for %zd bytes",
+                     storage_size);
```

Comment:
The idea here is to broadcast the potential error to the other ranks to avoid hanging in team creation. The error checking occurs during `ucc_tl_mlx5_a2a_init_progress` (line 301)
Should I remove this?

Reply:
I'm actually realizing that the broadcast only occurs within each node, so if one node fails, even with this mechanism, the other nodes will hang anyway.

So I can either remove all of this, or fix it by propagating the error down to the inter-node service allgather between node leaders